### PR TITLE
Engine - multi-party submissions [KVL-703]

### PR DESCRIPTION
**Pull request by kamil-da**
### Pull Request Checklist

This PR covers KVL-703.

This PR changes the API of `Engine` to support multi-party submissions.

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by kamil-da** [comment]
@remyhaemmerle-da @hurryabit @rautenrieth-da Currently this PR doesn't make any major changes to unit test for the `Engine` class - this should probably be changed.
Do you have any recommendations on what unit tests would be useful for this particular change? Should I mimic currently existing ones and create new tests but for multiple submitters instead of a single submitter?

File: ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala [hunk 1]
```diff
@@ -95,7 +95,7 @@ class Replay {
   def bench(): Unit = {
     val result = engine
       .replay(
-        benchmark.transaction.singleSubmitterOrThrow(),
+        Set(benchmark.transaction.singleSubmitterOrThrow()),
```
Comment (rautenrieth-da) [inline ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala:98]:
```suggestion
        benchmark.transaction.actAs.toSet,
```

`singleSubmitterOrThrow` is a placeholder used in all places that can't yet handle multiple submitters.

File: ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala [hunk 2]
```diff
@@ -130,7 +130,7 @@ class Replay {
     // before running the bench, we validate the transaction first to be sure everything is fine.
     val result = engine
       .validate(
-        benchmark.transaction.singleSubmitterOrThrow(),
+        Set(benchmark.transaction.singleSubmitterOrThrow()),
```
Comment (rautenrieth-da) [inline ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala:133]:
```suggestion
        benchmark.transaction.actAs.toSet,
```

**Review by rautenrieth-da — COMMENTED** [review]
> Do you have any recommendations on what unit tests would be useful for this particular change?

@remyhaemmerle-da is better suited to answer this, but I would not duplicate all tests. Maybe just add the following ones?
- One test that verifies that submitting a command with superfluous submitters works (i.e., copy a test that submits a command to create a contract, and add some extra submitters that are not used for anything)
- One test that executes a create command with multiple submitters that are all signatories (to demonstrate that multi-party submissions work in the engine)

**Comment by hurryabit** [comment]
@kamil-da Could you please ping @remyhaemmerle-da and me when this PR is ready for review by the language team.

File: ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala (hunk 1 above)
Comment (kamil-da) [reply ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala:98]: I might have misinterpreted the comment from the `singleSubmitterOrThrow()` method:
```
  // Note: this method will be removed when the entire kvutils code base
  // supports multi-party submissions
```
I thought that we have to still wait before removing that.
I will change the PR according to your comment.

**Comment by kamil-da** [comment]
@rautenrieth-da Thank you for the review. I agree with your proposal for unit tests 👍 
@hurryabit Sure, I will let you know when it's ready for review.

**Comment by kamil-da** [comment]
@rautenrieth-da Thank you for the review. I agree with your proposal for unit tests 👍 
@hurryabit Sure, I will let you know when it's ready for review.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 3]
```diff
@@ -161,7 +161,7 @@ class Engine(val config: EngineConfig = EngineConfig.Stable) {
     } yield (tx, meta)
 
   def replay(
-      submitter: Party,
+      submitters: Set[Party],
```
Comment (kamil-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:164]: @rautenrieth-da You've mentioned that we should use here `List[Party]` but `Set[Party]` seems to be used across the `Engine`, so I used the latter. Is that ok for you?

Comment (rautenrieth-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala:164]: Yes, IMO the interface should be consistent with the rest of `Engine`.

File: ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala (hunk 1 above)
Comment (kamil-da) [reply ledger/participant-state/kvutils/tools/engine-benchmark/src/benchmark/scala/ledger/participant/state/kvutils/tools/engine/benchmark/Replay.scala]: `actAs` was not available - I've used `submitters` instead.

**Review by remyhaemmerle-da — COMMENTED** [review]
Thinks looks pretty good to me. 

I have however a small favor to ask related to the mess preexisting in `EngineTest.scala`.

Could you duplicate the "minimal create command" instead of reusing it (in `EngineTest.scala` or in a separate file) to test muti-party submission independently and  run only the cases `"be translated"` , `"reinterpret to the same result"`, `"be validated"` ? 

Also I would like a test of replay showing failure when the submiters set are different at creation and replay. For instance you can test `multiPartyTemplate` with parties `Set(alice, bob, clara)` at creation and `Set(alice, bob)` at relplay and conversely.

**Comment by kamil-da** [comment]
@remyhaemmerle-da Thanks for the review!
Yes, I can duplicate and run only the cases you listed. I'll also implement the replay unit test. 👍

**Review by remyhaemmerle-da — APPROVED** [review]
Nice. Thanks a lot.

**Pull request merged**